### PR TITLE
FIR: resolve callable reference as lambda return

### PR DESCRIPTION
**Pull request by jsjeon**
The motivation is bb test `javaInterop/notNullAssertions/localEntities.kt`:
```kt
inline fun <T, U> expectAssertion(f: () -> (T) -> U): Unit? {
    try {
        F.passNull(f())
    } catch (e: NullPointerException) {
        return Unit
    }
    return null
}
fun box(): String {
    expectAssertion {
        fun f(x: String) {}
        ::f // unresolved reference for now
    } ?: return "FAIL: local function"
    ...
}
```
Investigation was filed as [KT-43949](https://youtrack.jetbrains.com/issue/KT-43949), but here is rephrased, more detailed explanation in a chronological order:

------

First, I double-checked if that's due to "local" entries, but even top-level function and its reference also fails.

Interestingly, if that last callable reference is stored to a local variable and then returned, everything works:
```kt
expectAssertion {
  fun local(...) { ... }
  val x = ::local // resolved
  x
}
```
That helped me find where things are different: last expression v.s. all the other expressions in the block are visited with different resolution modes. The former in a dependent resolution mode, whereas the latter in a given resolution mode.

So, my first trial was to call `resolveCallableReferenceWithSyntheticOuterCall` at `FirExpressionResolveTransformer#transformCallableReferenceAccess` no matter what. Well, it suffered from so many regressions, and the main cause was, when a callable reference was used as an argument, it would be rather resolved while resolving arguments, with better type info later.

My second trial was to pass a different resolution mode (in this case, lambda resolution mode is passed up to block resolution). However, that didn't work either, e.g., for lambda inside lambda, since inner lambda was visited in a lambda resolution mode with incorrect expected type (outer lambda's expected type).

Note that, in the original example, the lambda's type variables won't be resolved until the body is visited/resolved. Due to the lack of the expected return type of the lambda, it seemed inevitable to visit the callable reference with a dependent mode. With that, that reference is not resolved, then the lambda's return type is still not resolved. So, revisiting the lambda body again isn't a solution.

------

Luckily, one of regressions gave me another insight, which is filed as [KT-43948](https://youtrack.jetbrains.com/issue/KT-43948). While trying to tweak the resolution mode, I found a case like this:
```kt
    x.foo<Int> {
        if (x.hashCode() == 0) return@foo <!UNRESOLVED_REFERENCE!>::bar<!>
        ::bar // became unresolved!
    }
```
Those are the same callable references, but only the former was regarded as unresolved. The latter became unresolved if I force resolving it. That means, as a last expression of a lambda, its resolution trial has been avoided. On the other hand, the former is visited in an independent mode, as part of jump expression.

With further investigation, I found:
```kt
fun main() {
    val x = run { ::run } // no error
}
```
which isn't shown as an error, but the compilation fails, because `::run` is not resolved actually. The issue is that unresolved reference is hidden as a form of `FirNamedReference` without a candidate symbol.

I double-checked that the original example works if it's changed like this:
```kt
expectAssertion {
  fun local(...) { ... }
  return@expectAssertion ::local
}
```

So, here is the idea. We have a step to add `return` to the lambda if that lambda doesn't return `Nothing` or `Unit`. Callable reference is not the case, of course, since its type would be a functional type or `KProperty` (or something else?). That step is conducted _after_ lambda resolution, but it would be too late in this case, since unresolved reference means unresolved return type of the lambda, which blocks `return` insertion.

Instead, we can add `return` to the lambda _before_ lambda resolution if the last expression is about to not return `Nothing` or `Unit`. That will make a callable reference as lambda return be resolved during lambda resolution, making the lambda return type resolved too.

------

To show the aforementioned idea (implemented as 2nd commit) doesn't introduce any regressions, 1st commit exposes hidden unresolved reference ahead(?).

### CODE REVIEW

File: compiler/testData/diagnostics/tests/callableReference/kt32267.kt [hunk 1]
```diff
@@ -1,4 +1,3 @@
-// FIR_IDENTICAL
```
Comment (mglukhikh) [inline compiler/testData/diagnostics/tests/callableReference/kt32267.kt:1]: At least this change looks very suspicious

Comment (jsjeon) [reply compiler/testData/diagnostics/tests/callableReference/kt32267.kt:1]: Not really, we just expose hidden, unresolved references. That is all about https://youtrack.jetbrains.com/issue/KT-43948. See my other comments in the newly added .fir.kr (slightly right upper side).

Also, if you see the 1st commit separately, there are more cases that we drop `// FIR_IDENTICAL`. Luckily, all the others were recovered by the 2nd commit.

File: compiler/testData/diagnostics/tests/callableReference/kt32267.fir.kt [hunk 2]
```diff
@@ -0,0 +1,39 @@
+// SKIP_TXT
+// !DIAGNOSTICS: -UNUSED_PARAMETER
+// !LANGUAGE: +NewInference
+
+fun main() {
+    Configuration().commands {
+        Command1 { <!UNRESOLVED_REFERENCE!>someService::execute<!> } // Overload resolution ambiguity. All these functions match.
+        Command2 { <!UNRESOLVED_REFERENCE!>someService::execute<!> } // Overload resolution ambiguity. All these functions match.
```
Comment (jsjeon) [inline compiler/testData/diagnostics/tests/callableReference/kt32267.fir.kt:8]: So, these two break `// FIR_IDENTICAL`, and indeed, these are `FirNamedReference` without pointing any target/candidate symbol. This will fail eventually if you compile this.

File: compiler/testData/diagnostics/tests/callableReference/kt32267.kt (hunk 1 above)
Comment (mglukhikh) [reply compiler/testData/diagnostics/tests/callableReference/kt32267.kt:1]: Probably you're right, yes.

**Comment by jsjeon** [comment]
Friendly ping.

**Comment by dzharkov** [comment]
> Friendly ping.

Sorry, I'm on my vacation, so I won't be able to answer until 11th Jan

**Comment by jsjeon** [comment]
> > Friendly ping.
> 
> Sorry, I'm on my vacation, so I won't be able to answer until 11th Jan

No problem! I just wondered if this is not under your radar. Enjoy the vacation!

**Comment by jsjeon** [comment]
While rebasing and running tests again, I found that relocated tests at https://github.com/JetBrains/kotlin/commit/39e579db91c12921d412365b71259489ee50725f were failing, and discovered that neither of `FirCallCompletionResultsWriterTransformer#transformCallableReferenceAccess` (that I exposed referenced without candidate as unresolved reference) or `FirExpressionsResolveTransformer#transformCallableReferenceAccess` is a good place to determine if the reference without candidates is indeed unresolved or not for this case:
```kt
fun poll41(): Any? {
    val inv = try { ::bar2 } finally { ::foo2 }
    return inv()
}
```
where `bar2` and `foo2` return type mismatch (but could be approximated to their super type). The former is resolved at expression resolution, whereas the latter is resolved by postponed argument analyzer. So, I dropped the 1st commit and let 
[KT-43948](https://youtrack.jetbrains.com/issue/KT-43948) open as-is.

I still believe that simply named reference without candidates or target symbol should be treated as unresolved reference at the end of resolution or before `fir2ir` conversion. Also, I think the idea of adding `return` as if the last expression of the lambda won't return `Nothing` or `Unit` is still valid. But, let me know if you think there is a better way to deal with this situation.

**Review by dzharkov — COMMENTED** [review]
Sorry for the delay, again :(
I've dug into the problem and still didn't manage to finish my investigation.

The reason I started digging in is that it sounded suspicious to me that we need other returns to make it work
Because we obtain return expressions from `FirDataFlowAnalyzer#returnExpressionsOfAnonymousFunction` at `FirCallCompleter.LambdaAnalyzerImpl#analyzeAndGetLambdaReturnArguments`
and the last expression even without explicit return should be considered as a return-argument in terms of data flow graph.

Then, I figured out that we just don't add nodes for callable reference to the graph and my first attempt was https://github.com/JetBrains/kotlin/commit/147b555992ffcea309918b7f4653655634f206f9

It actually fixes most of the tests from the PR, but I'm also started to avoid that kind of regressions like `kt32267.fir.kt`

And here I am at https://github.com/JetBrains/kotlin/compare/rr/zharkov-fir-callable-references-9
Most of the tests are fixed, but new ones start failing for obscure reasons.

So, basically, current PR seems to be fixing some tests but looks like it does that indirectly disguising some problems like the absence of callable reference nodes.
Thus, as I hope to finish my attempt, unfortunately, this PR might become unnecessary.
But currently, I'd leave the PR open in case we'll figure out I was wrong and your solution is much simpler than mine.

Again, I'm really sorry about not giving more feedback right at the time, but I've been a bit passionate about the problem, and each time I was sure that I'll manage to understand what's going on.
I will try to warn if something like that happens next time.

**Comment by jsjeon** [comment]
Thanks for the update! I've moved on to checkers, so I wasn't blocked technically, no worries.

Good to know that such callable reference as a lambda return should be a return-argument in DFA.

**Comment by dzharkov** [comment]
I've pushed my changes, thus all problems have been resolved
